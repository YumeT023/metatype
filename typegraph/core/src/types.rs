// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use enum_dispatch::enum_dispatch;

use crate::core::{TypeBase, TypeFunc, TypeId, TypeInteger, TypeProxy, TypeStruct};

#[allow(clippy::derivable_impls)]
impl Default for TypeBase {
    fn default() -> Self {
        Self { name: None }
    }
}

#[derive(Debug)]
pub struct Proxy(pub TypeProxy);

#[derive(Debug)]
pub struct Struct(pub TypeBase, pub TypeStruct);

#[derive(Debug)]
pub struct Integer(pub TypeBase, pub TypeInteger);

#[derive(Debug)]
pub struct Func(pub TypeBase, pub TypeFunc);

#[derive(Debug)]
#[enum_dispatch(TypeFun)]
pub enum T {
    Proxy(Proxy),
    Struct(Struct),
    Integer(Integer),
    Func(Func),
}

#[enum_dispatch]
pub trait TypeFun {
    fn get_repr(&self, id: TypeId) -> String;
    fn get_base(&self) -> Option<&TypeBase>;
}

impl TypeFun for Proxy {
    fn get_repr(&self, id: TypeId) -> String {
        format!("proxy(#{id})")
    }

    fn get_base(&self) -> Option<&TypeBase> {
        None
    }
}

impl TypeFun for Integer {
    fn get_repr(&self, id: TypeId) -> String {
        let c = self.1;
        let data = [
            Some(format!("#{id}")),
            c.min.map(|min| format!("min={min}")),
            c.max.map(|max| format!("max={max}")),
        ]
        .into_iter()
        .flatten()
        .collect::<Vec<_>>()
        .join(", ");
        format!("integer({data})")
    }

    fn get_base(&self) -> Option<&TypeBase> {
        Some(&self.0)
    }
}

impl TypeFun for Struct {
    fn get_repr(&self, id: TypeId) -> String {
        let c = &self.1;
        let props = c
            .props
            .iter()
            .map(|(name, tpe_id)| format!("[{name}] => #{tpe_id}"))
            .collect::<Vec<_>>()
            .join(", ");
        format!("struct(#{id}, {props})")
    }

    fn get_base(&self) -> Option<&TypeBase> {
        Some(&self.0)
    }
}

impl TypeFun for Func {
    fn get_repr(&self, id: TypeId) -> String {
        let c = &self.1;
        format!("func(#{id}, #{} => #{})", c.inp, c.out)
    }

    fn get_base(&self) -> Option<&TypeBase> {
        Some(&self.0)
    }
}
