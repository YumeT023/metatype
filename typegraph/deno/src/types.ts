import { core } from "../gen/typegraph_core.js";
import type { IntegerConstraints } from "../gen/exports/core.d.ts";
import { NullableOptional } from "./utils/type_utils.ts";

class Tpe {
  constructor(public id: number) {}

  get repr(): string | null {
    return core.getTypeRepr(this.id);
  }

  asInteger(): Integer {
    if (this instanceof Integer) {
      return this;
    }

    const typeData = core.typeAsInteger(this.id);
    if (typeData != null) {
      return new Integer(this.id, typeData);
    }
    throw new Error("Not an integer");
  }

  asStruct(): Struct<Record<string, Tpe>> {
    if (this instanceof Struct) {
      return this;
    }

    const typeData = core.typeAsStruct(this.id);
    if (typeData != null) {
      return new Struct(this.id, {
        ...typeData,
        props: Object.fromEntries(
          typeData.props.map(([name, id]) => [name, new Tpe(id)]),
        ),
      });
    }
    throw new Error("Not a struct");
  }

  asTpe(): Tpe {
    return new Tpe(this.id);
  }
}

class Integer extends Tpe implements Readonly<IntegerConstraints> {
  readonly min?: number;
  readonly max?: number;

  constructor(id: number, data: IntegerConstraints) {
    super(id);
    this.min = data.min;
    this.max = data.max;
  }
}

export function integer(data: NullableOptional<IntegerConstraints> = {}) {
  return new Integer(core.integerb(data).id, data);
}

class Struct<P extends { [key: string]: Tpe }> extends Tpe {
  props: P;
  constructor(id: number, { props }: { props: P }) {
    super(id);
    this.props = props;
  }
}

export function struct<P extends { [key: string]: Tpe }>(props: P): Struct<P> {
  return new Struct(
    core.structb({
      props: Object.entries(props).map(([name, typ]) => [name, typ.id]),
    }).id,
    {
      props,
    },
  );
}
