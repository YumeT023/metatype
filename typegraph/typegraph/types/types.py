# Copyright Metatype under the Elastic License 2.0.
from copy import deepcopy
import os
from types import NoneType
from typing import Any
from typing import Dict
from typing import get_args
from typing import get_origin
from typing import List
from typing import Optional
from typing import Set
from typing import Tuple
from typing import Type
from typing import Union

from frozendict import frozendict
import orjson
from typegraph.graphs.builder import Collector
from typegraph.graphs.node import Node
from typegraph.graphs.typegraph import NodeProxy
from typegraph.graphs.typegraph import TypeGraph
from typegraph.graphs.typegraph import TypegraphContext
from typegraph.materializers.base import Materializer
from typegraph.materializers.base import Runtime
from typegraph.policies import Policy


if os.environ.get("DEBUG"):
    import debugpy

    debugpy.listen(5678)
    print("Waiting for debugger attach...")
    debugpy.wait_for_client()


def replace(obj, **kwargs):
    # handle errors
    fields = {k: deepcopy(v) for k, v in vars(obj).items() if k not in kwargs}
    fields.update(**kwargs)
    return type(obj)(**fields)


def remove_none_values(obj):
    return {k: v for k, v in obj.items() if v is not None}


def is_optional(tpe: Type):
    # Optional = Union[T, NoneType]
    return get_origin(tpe) is Union and NoneType in get_args(tpe)


class Secret(Node):
    secret: str

    def __init__(self, secret: str):
        super().__init__("secrets")
        self.secret = secret

    def data(self, collector: "Collector") -> dict:
        return self.secret  # this is a string


class typedef(Node):
    graph: TypeGraph
    name: str
    description: Optional[str]
    runtime: Optional["Runtime"]
    inject: Optional[Union[str, "typedef"]]
    injection: Optional[Any]
    policies: Tuple[Policy, ...]

    def __init__(
        self,
        graph: Optional[TypeGraph] = None,
        runtime: Optional["Runtime"] = None,
        name: Optional[str] = None,
        description: Optional[str] = None,
        inject: Optional[Union[str, "typedef"]] = None,
        injection: Optional[Any] = None,
        policies: Optional[Tuple[Policy, ...]] = None,
        **kwargs,
    ):
        super().__init__(
            kwargs["collector_target"]
            if "collector_target" in kwargs
            else Collector.types
        )

        for k, v in type(self).__annotations__.items():
            # manage dataclass like behaviour (e.g. field) and block bad instanciation (e.g. [])
            if is_optional(v):
                setattr(self, k, kwargs.pop(k, getattr(self, k)))

        if graph is None:
            active_graph = TypegraphContext.get_active()
            if active_graph is None:
                raise Exception("no typegraph context")
            self.graph = active_graph
            self.graph.register(self)  # FIXME still needed?
        else:
            self.graph = graph

        # TODO ensure subclass type
        self.name = f"{self.type}_{self.graph.next_type_id()}" if name is None else name
        self.description = description
        self.runtime = runtime
        self.inject = inject
        self.injection = None if self.inject is None else injection
        self.policies = tuple() if policies is None else policies

    @property
    def type(self):
        return type(self).__name__

    @property
    def edges(self) -> List[Node]:
        secret = self.inject if self.injection == "secret" else None
        return (
            super().edges
            + list(self.policies)
            + list(filter(None, [self.runtime, secret]))
        )

    def named(self, name: str) -> "typedef":
        # TODO: ensure name is not already used in typegraph
        return replace(self, name=name)

    def describe(self, description: str) -> "typedef":
        return replace(self, description=description)

    def within(self, runtime):
        if runtime is None:
            raise Exception(f"cannot set runtime to None")

        if self.runtime is not None and self._runtime != runtime:
            raise Exception(
                f"trying to set a different runtime {runtime}, already is {self.runtime}"
            )

        self.runtime = runtime

        return self

    def _propagate_runtime(self, runtime: "Runtime", visited: Set["typedef"] = None):
        if visited is None:
            visited = set()
        elif self in visited:
            return
        else:
            visited.add(self)

        if self.runtime is None:
            self.runtime = runtime

        for e in self.edges:
            if isinstance(e, typedef):
                e._propagate_runtime(self.runtime, visited)

    def set(self, value):
        if self.inject is not None:
            raise Exception(f"{self.name} can only have one injection")

        return replace(self, injection="raw", inject=orjson.dumps(value).decode())

    def from_secret(self, secret_name):
        if self.inject is not None:
            raise Exception(f"{self.name} can only have one injection")

        return replace(self, injection="secret", inject=Secret(secret_name))

    def from_parent(self, sibiling: "NodeProxy"):
        # TODO: check for same type and value in same context
        if self.inject is not None:
            raise Exception(f"{self.name} can only have one injection")

        return replace(self, injection="parent", inject=sibiling)

    def from_context(self, claim: str):
        if self.inject is not None:
            raise Exception(f"{self.name} can only have one injection")

        return replace(self, injection="context", inject=claim)

    def add_policy(
        self,
        *policies: List["Policy"],
    ):
        return replace(self, policies=self.policies + policies)

    def data(self, collector) -> dict:
        ret = {
            "type": self.type,
            "title": self.name,
            "runtime": collector.collect(self.runtime),
            "policies": [collector.collect(p) for p in self.policies],
        }
        if self.description is not None:
            ret["description"] = self.description

        if self.inject is not None:
            ret["injection"] = self.injection
            if self.injection == "parent":
                ret["inject"] = collector.collect(self.inject)
            elif self.injection == "secret":
                ret["inject"] = self.inject.secret
            else:
                ret["inject"] = self.inject

        return ret

    def optional(self, default_value: Optional[Any] = None):
        if isinstance(self, optional):
            return self

        with self.graph:
            return optional(self, default_value=default_value)

    def __repr__(self) -> str:
        def pretty(v):
            if isinstance(v, frozendict):
                return dict(v)
            return v

        attrs = ", ".join(
            f"{k} â†’ {pretty(v)}"
            for k, v in vars(self).items()
            if k not in ("graph",) and v is not None
        )
        return f"{self.type}({attrs})"


class optional(typedef):
    of: typedef
    default_value: Optional[Any] = None

    def __init__(self, of: typedef, **kwargs):
        super().__init__(**kwargs)
        self.of = of

    def default(self, value):
        if self.default_value is not None:
            raise Exception(f"{self.name} can only have one injection")

        return replace(self, default_value=value)

    @property
    def edges(self) -> List[Node]:
        return super().edges + [self.of]

    def data(self, collector) -> dict:
        return {
            **super().data(collector),
            "item": collector.collect(self.of),
            "default_value": self.default_value,
        }


class boolean(typedef):
    pass


class number(typedef):
    _min: Optional[float] = None
    _max: Optional[float] = None
    _x_min: Optional[float] = None
    _x_max: Optional[float] = None
    _multiple_of: Optional[float] = None

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def min(self, x: float) -> "number":
        return replace(self, _min=x)

    def max(self, x: float) -> "number":
        return replace(self, _max=x)

    def x_min(self, x: float) -> "number":
        return replace(self, _x_min=x)

    def x_max(self, x: float) -> "number":
        return replace(self, _x_max=x)

    def multiple_of(self, x: float) -> "number":
        return replace(self, _multiple_of=x)

    def data(self, collector) -> dict:
        return {
            **super().data(collector),
            **remove_none_values(
                {
                    "minimum": self._min,
                    "maximum": self._max,
                    "exclusiveMinimum": self._x_min,
                    "exclusiveMaximum": self._x_max,
                    "multipleOf": self._multiple_of,
                }
            ),
        }


class integer(number):
    pass


class string(typedef):
    _min: Optional[int] = None
    _max: Optional[int] = None
    _pattern: Optional[str] = None
    _format: Optional[str] = None

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def min(self, n: int) -> "string":
        return replace(self, _min=n)

    def max(self, n: int) -> "string":
        return replace(self, _max=n)

    def pattern(self, p: str) -> "string":
        return replace(self, _pattern=p)

    def format(self, f: str) -> "string":
        return replace(self, _format=f)

    def uuid(self) -> "string":
        return self.format("uuid")

    def email(self) -> "string":
        return self.format("email")

    def uri(self) -> "string":
        return self.format("uri")

    def hostname(self) -> "string":
        return self.format("hostname")

    def data(self, collector) -> dict:
        return {
            **super().data(collector),
            **remove_none_values(
                {
                    "minLength": self._min,
                    "maxLength": self._max,
                    "pattern": self._pattern,
                    "format": self._format,
                }
            ),
        }


def uuid() -> string:
    return string().uuid()


def email() -> string:
    return string().email()


def uri() -> string:
    return string().uri()


class struct(typedef):
    props: Dict[str, typedef]
    required: List[str]

    def __init__(self, props=None, **kwargs):
        super().__init__(**kwargs)
        self.props = dict()

        if "required" in kwargs:  # re constructing from data
            if props is None:
                raise Exception("invalid")
            self.required = kwargs["required"]
            self.props = props

        else:
            self.required = []

            for k, v in props.items() if props is not None else []:
                if isinstance(v, optional):
                    v = v.of
                else:
                    self.required.append(k)
                self.props[k] = v

    @property
    def type(self):
        return "object"

    @property
    def edges(self) -> List[Node]:
        return super().edges + list(self.props.values())

    def data(self, collector) -> dict:
        return {
            **super().data(collector),
            "properties": {k: collector.collect(v) for k, v in self.props.items()},
            "required": self.required,
        }


class array(typedef):
    of: typedef

    _min: Optional[int] = None
    _max: Optional[int] = None
    _unique_items: Optional[bool] = None
    # _min_contains: Optional[int] = None
    # _max_contains: Optional[int] = None

    def __init__(self, of: typedef, **kwargs):
        super().__init__(**kwargs)
        self.of = of

    def min(self, n: int) -> "array":
        return replace(self, _min=n)

    def max(self, n: int) -> "array":
        return replace(self, _max=n)

    def unique_items(self, b: bool) -> "array":
        return replace(self, _unique_items=b)

    @property
    def edges(self) -> List[Node]:
        return super().edges + [self.of]

    def data(self, collector) -> dict:
        return {
            **super().data(collector),
            "items": collector.collect(self.of),
            **remove_none_values(
                {
                    "minItems": self._min,
                    "maxItems": self._max,
                    "uniqueItems": self._unique_items,
                }
            ),
        }


class func(typedef):
    inp: struct
    out: typedef
    mat: Materializer

    # if not safe, output will be typechecked
    safe: bool

    rate_calls: bool
    rate_weight: Optional[int] = None

    def __init__(
        self,
        inp: struct,
        out: typedef,
        mat: Materializer,
        safe=True,
        rate_calls=False,
        rate_weight=None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.inp = inp
        self.out = out
        self.mat = mat
        self.safe = safe

        self.runtime = mat.runtime

        self.rate_calls = rate_calls
        self.rate_weight = rate_weight

    def rate(self, weight=None, calls=None):
        return replace(self, rate_weight=weight, rate_calls=calls)

    @property
    def edges(self) -> List[Node]:
        return super().edges + [self.inp, self.out, self.mat]

    @property
    def type(self) -> str:
        return "function"

    def data(self, collector) -> dict:
        return {
            **super().data(collector),
            "input": collector.collect(self.inp),
            "output": collector.collect(self.out),
            "materializer": collector.collect(self.mat),
            "rate_weight": self.rate_weight,
            "rate_calls": self.rate_calls,
        }

    # def compose(self, other: "func") -> "func":
    #     assert self.out == other.inp
    #     # what if other == gen?
    #     return func(self, other, IdentityMat())

    # def __mul__(self, other: "func") -> "func":
    #     return self.compose(other)


def gen(out: typedef, mat: Materializer, **kwargs) -> func:
    return func(struct(), out, mat, **kwargs)
