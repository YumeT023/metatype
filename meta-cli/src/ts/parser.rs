// Copyright Metatype OÃœ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.

use anyhow::bail;
use anyhow::Result;
use std::collections::HashSet;
use std::path::Path;
use swc_common::errors::{ColorConfig, Handler};
use swc_common::hygiene::Mark;
use swc_common::sync::Lrc;
use swc_common::FileName;
use swc_common::Globals;
use swc_common::SourceMap;
use swc_common::DUMMY_SP;
use swc_common::GLOBALS;
use swc_ecma_ast::*;
use swc_ecma_transforms::typescript::strip::strip;
use swc_ecmascript::parser::parse_file_as_module;
use swc_ecmascript::parser::parse_file_as_script;
use swc_ecmascript::{
    codegen,
    codegen::{text_writer::JsWriter, Emitter},
    parser::Syntax,
    visit::{Fold, FoldWith},
};

fn named_import(tpe: String, alias: String) -> ImportNamedSpecifier {
    ImportNamedSpecifier {
        local: Ident::new(tpe.into(), DUMMY_SP),
        span: DUMMY_SP,
        imported: Some(ModuleExportName::Ident(Ident::new(alias.into(), DUMMY_SP))),
        is_type_only: false,
    }
}

fn import(specifier: ImportNamedSpecifier, source: &str) -> ImportDecl {
    ImportDecl {
        span: DUMMY_SP,
        specifiers: vec![ImportSpecifier::Named(specifier)],
        src: Box::new(Str {
            span: DUMMY_SP,
            value: source.into(),
            raw: Default::default(),
        }),
        type_only: true,
        asserts: None,
    }
}

fn tpe(name: String) -> TsTypeAnn {
    TsTypeAnn {
        span: DUMMY_SP,
        type_ann: Box::new(TsType::TsTypeRef(TsTypeRef {
            span: DUMMY_SP,
            type_name: TsEntityName::Ident(Ident::new(name.into(), DUMMY_SP)),
            type_params: None,
        })),
    }
}

fn export_function(name: String, inp: String, out: String, body: Option<BlockStmt>) -> ExportDecl {
    let f = FnDecl {
        ident: Ident::new(name.into(), DUMMY_SP),
        declare: false,
        function: Box::new(Function {
            params: vec![Param {
                span: DUMMY_SP,
                decorators: vec![],
                pat: Pat::Ident(BindingIdent {
                    id: Ident::new("arg".into(), DUMMY_SP),
                    type_ann: Some(Box::new(tpe(inp))),
                }),
            }],
            decorators: vec![],
            span: DUMMY_SP,
            body,
            is_generator: false,
            is_async: false,
            type_params: None,
            return_type: Some(Box::new(tpe(out))),
        }),
    };
    ExportDecl {
        decl: Decl::Fn(f),
        span: DUMMY_SP,
    }
}

pub struct MyVisitor {
    pub source: String,
    imports: Vec<(String, String)>,
}

impl Fold for MyVisitor {
    fn fold_module(&mut self, n: Module) -> Module {
        let mut rewrite = n
            .body
            .into_iter()
            .filter(|s| match &s {
                ModuleItem::ModuleDecl(ModuleDecl::Import(imp)) => {
                    !(imp.src.value.to_string() == "./output.ts" && imp.type_only)
                }
                _ => true,
            })
            .collect::<Vec<_>>();

        if let Some(idx) = rewrite.iter().position(|s| match &s {
            ModuleItem::ModuleDecl(ModuleDecl::ExportDecl(ExportDecl {
                decl: Decl::Fn(f),
                ..
            })) => f.ident.sym.to_string() == "apply",
            _ => false,
        }) {
            if let ModuleItem::ModuleDecl(ModuleDecl::ExportDecl(ExportDecl {
                decl: Decl::Fn(f),
                span: _,
            })) = rewrite.swap_remove(idx)
            {
                let new_exp = export_function(
                    "testaaa".to_string(),
                    "Input".to_string(),
                    "Output".to_string(),
                    f.function.body,
                );
                rewrite.push(ModuleItem::ModuleDecl(ModuleDecl::ExportDecl(new_exp)));
            }
        } else {
            let new_exp = export_function(
                "testaaa".to_string(),
                "Input".to_string(),
                "Output".to_string(),
                None,
            );
            rewrite.push(ModuleItem::ModuleDecl(ModuleDecl::ExportDecl(new_exp)));
        }

        for imp in &self.imports {
            let imp = import(named_import(imp.0.clone(), imp.1.clone()), "./output.ts");
            rewrite.push(ModuleItem::ModuleDecl(ModuleDecl::Import(imp)));
        }
        rewrite.rotate_right(self.imports.len());
        Module { body: rewrite, ..n }
    }
    fn fold_export_decl(&mut self, n: ExportDecl) -> ExportDecl {
        match &n.decl {
            Decl::Fn(FnDecl {
                ident,
                declare: _,
                function,
            }) if ident.sym.to_string() == "apply" => {
                println!("TEST {:?}", function.params);
            }
            _ => (),
        }
        n
    }
}

pub fn parse_module<P>(mod_path: P) -> Result<Module>
where
    P: AsRef<Path>,
{
    let cm: Lrc<SourceMap> = Default::default();
    let handler = Handler::with_tty_emitter(ColorConfig::Auto, true, false, Some(cm.clone()));

    let fm = cm.load_file(mod_path.as_ref())?;
    // .expect(&format!("failed to load file: {:?}", mod_path.as_ref()));

    let mut errors = vec![];
    parse_file_as_module(
        &fm,
        Syntax::Typescript(Default::default()),
        EsVersion::latest(),
        None,
        &mut errors,
    )
    .or_else(|e| {
        e.into_diagnostic(&handler).emit();
        bail!("could not compile module")
    })
}

pub fn parse_module_source(source: String) -> Result<Module> {
    let cm: Lrc<SourceMap> = Default::default();
    let handler = Handler::with_tty_emitter(ColorConfig::Auto, true, false, Some(cm.clone()));

    let fm = cm.new_source_file(FileName::Anon, source);

    let mut errors = vec![];
    parse_file_as_module(
        &fm,
        Syntax::Typescript(Default::default()),
        EsVersion::latest(),
        None,
        &mut errors,
    )
    .or_else(|e| {
        e.into_diagnostic(&handler).emit();
        bail!("could not compile module")
    })
}

pub fn transform_module(module: Module) -> Result<String> {
    let globals = Globals::default();
    let module = GLOBALS.set(&globals, || {
        let top_level_mark = Mark::new();
        module.fold_with(&mut strip(top_level_mark))
    });

    with_emitter(|emitter| {
        emitter.emit_module(&module).map_or_else(
            |e| bail!("error while emitting module code: {e}"),
            |_| Ok(()),
        )
    })
}

pub fn transform_script(source: String) -> Result<String> {
    let cm: Lrc<SourceMap> = Default::default();
    let handler = Handler::with_tty_emitter(ColorConfig::Auto, true, false, Some(cm.clone()));
    let fm = cm.new_source_file(FileName::Anon, source);

    let mut errors = vec![];
    let script = parse_file_as_script(
        &fm,
        Syntax::Typescript(Default::default()),
        EsVersion::latest(),
        None,
        &mut errors,
    )
    .or_else(|e| {
        e.into_diagnostic(&handler).emit();
        bail!("could not compile script")
    })?;

    for e in errors {
        e.into_diagnostic(&handler).emit();
    }

    let globals = Globals::default();
    let script = GLOBALS.set(&globals, || {
        let top_level_mark = Mark::new();
        script.fold_with(&mut strip(top_level_mark))
    });

    with_emitter(|emitter| {
        emitter.emit_script(&script).map_or_else(
            |e| bail!("error while emitting script code: {e}"),
            |_| Ok(()),
        )
    })
}

pub fn get_exported_functions(mod_body: &Vec<ModuleItem>) -> HashSet<String> {
    let mut res = HashSet::default();
    for mod_item in mod_body {
        match mod_item {
            ModuleItem::ModuleDecl(ModuleDecl::ExportDecl(decl)) => {
                if let Decl::Fn(fn_decl) = &decl.decl {
                    assert!(res.insert(fn_decl.ident.sym.to_string()));
                }
            }
            ModuleItem::ModuleDecl(ModuleDecl::ExportDefaultDecl(decl)) => {
                if let DefaultDecl::Fn(_) = &decl.decl {
                    assert!(res.insert("default".to_string()));
                }
            }
            _ => (),
        }
    }

    res
}

fn with_emitter<'a, C>(emit: C) -> Result<String>
where
    C: FnOnce(&mut Emitter<'a, JsWriter<&mut Vec<u8>>, SourceMap>) -> Result<()>,
{
    let cm: Lrc<SourceMap> = Default::default();
    let mut buf = vec![];
    let mut emitter = Emitter {
        cfg: codegen::Config {
            target: EsVersion::latest(),
            ascii_only: false,
            minify: true,
            omit_last_semi: false,
        },
        cm: cm.clone(),
        comments: None,
        wr: JsWriter::new(cm, "", &mut buf, None),
    };

    emit(&mut emitter)?;

    Ok(String::from_utf8_lossy(&buf).to_string())
}
