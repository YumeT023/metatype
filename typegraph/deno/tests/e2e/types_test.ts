import { assertEquals } from "std/testing/asserts.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { dirname, fromFileUrl, join, resolve } from "std/path/mod.ts";
import * as t from "../../src/types.ts";

const dirName = dirname(fromFileUrl(import.meta.url));

async function serialize(path: string): Promise<string> {
  const cmd = new Deno.Command(
    "cargo",
    {
      args: [
        "run",
        "-p",
        "meta-cli",
        "-F",
        "typegraph-next",
        "-q",
        "--",
        "serialize",
        "--pretty",
        "-f",
        resolve(dirName, path),
      ],
      stdout: "piped",
      stderr: "inherit",
    },
  );

  const out = await cmd.output();
  if (out.success) {
    const res = new TextDecoder().decode(out.stdout);
    return res;
  }
  throw new Error("Error");
}

Deno.test("types", async (t) => {
  assertSnapshot(t, await serialize("types.tg.ts"));
});
