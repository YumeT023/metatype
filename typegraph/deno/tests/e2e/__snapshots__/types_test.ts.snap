export const snapshot = {};

snapshot[`types 1`] = `
"[
" +
  "  {
" +
  '    "\$id": "https://metatype.dev/specs/0.0.1.json",
' +
  '    "types": [
' +
  "      {
" +
  '        "type": "object",
' +
  '        "title": "test-types",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {},
' +
  '        "properties": {
' +
  '          "one": 1,
' +
  '          "two": 5
' +
  "        },
" +
  '        "required": [
' +
  '          "one",
' +
  '          "two"
' +
  "        ]
" +
  "      },
" +
  "      {
" +
  '        "type": "function",
' +
  '        "title": "func_10",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {},
' +
  '        "input": 2,
' +
  '        "output": 3,
' +
  '        "materializer": 0,
' +
  '        "rate_weight": null,
' +
  '        "rate_calls": false
' +
  "      },
" +
  "      {
" +
  '        "type": "object",
' +
  '        "title": "object_4",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {},
' +
  '        "properties": {
' +
  '          "a": 3,
' +
  '          "b": 4
' +
  "        },
" +
  '        "required": []
' +
  "      },
" +
  "      {
" +
  '        "type": "integer",
' +
  '        "title": "integer_0",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {}
' +
  "      },
" +
  "      {
" +
  '        "type": "integer",
' +
  '        "title": "integer_3",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {}
' +
  "      },
" +
  "      {
" +
  '        "type": "function",
' +
  '        "title": "func_11",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {},
' +
  '        "input": 6,
' +
  '        "output": 8,
' +
  '        "materializer": 0,
' +
  '        "rate_weight": null,
' +
  '        "rate_calls": false
' +
  "      },
" +
  "      {
" +
  '        "type": "object",
' +
  '        "title": "object_7",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {},
' +
  '        "properties": {
' +
  '          "id": 7,
' +
  '          "post": 8
' +
  "        },
" +
  '        "required": []
' +
  "      },
" +
  "      {
" +
  '        "type": "integer",
' +
  '        "title": "integer_5",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {}
' +
  "      },
" +
  "      {
" +
  '        "type": "object",
' +
  '        "title": "object_9",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {},
' +
  '        "properties": {
' +
  '          "id": 9,
' +
  '          "author": 6
' +
  "        },
" +
  '        "required": []
' +
  "      },
" +
  "      {
" +
  '        "type": "integer",
' +
  '        "title": "integer_8",
' +
  '        "runtime": 0,
' +
  '        "policies": [],
' +
  '        "config": {}
' +
  "      }
" +
  "    ],
" +
  '    "materializers": [],
' +
  '    "runtimes": [],
' +
  '    "policies": [],
' +
  '    "meta": {
' +
  '      "secrets": [],
' +
  '      "cors": {
' +
  '        "allow_origin": [],
' +
  '        "allow_headers": [],
' +
  '        "expose_headers": [],
' +
  '        "allow_methods": [],
' +
  '        "allow_credentials": false,
' +
  '        "max_age_sec": null
' +
  "      },
" +
  '      "auths": [],
' +
  '      "rate": null,
' +
  '      "version": "0.0.1"
' +
  "    }
" +
  "  }
" +
  "]"
`;
